File: packages/core/src/watch-polling.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  fsTypeNeedsPolling,
  parseMountFsType,
  parsePollingOverride,
  hasRemoteEnvIndicator,
  decidePolling,
  pollingInterval,
  shouldUsePolling,
  withAuthoritativeChokidarEnv,
} from "./watch-polling.js";

// --- fsTypeNeedsPolling ---

test("fsTypeNeedsPolling: 網路 / 虛擬掛載需要 polling", () => {
  for (const t of ["9p", "v9fs", "drvfs", "cifs", "smb3", "nfs", "nfs4", "vboxsf", "prl_fs"]) {
    assert.equal(fsTypeNeedsPolling(t), true, t);
  }
});

test("fsTypeNeedsPolling: fuse.* 一律需要 polling（含 bare fuse / fuseblk）", () => {
  assert.equal(fsTypeNeedsPolling("fuse.sshfs"), true);
  assert.equal(fsTypeNeedsPolling("fuse"), true);
  assert.equal(fsTypeNeedsPolling("fuseblk"), true);
});

test("fsTypeNeedsPolling: 原生檔案系統不需要 polling", () => {
  for (const t of ["ext4", "overlay", "btrfs", "xfs", "tmpfs"]) {
    assert.equal(fsTypeNeedsPolling(t), false, t);
  }
});

test("fsTypeNeedsPolling: 大小寫不敏感、null 回 false", () => {
  assert.equal(fsTypeNeedsPolling("9P"), true);
  assert.equal(fsTypeNeedsPolling("NFS4"), true);
  assert.equal(fsTypeNeedsPolling(null), false);
});

// --- parseMountFsType ---

const MOUNTS = [
  "rootfs / rootfs rw 0 0",
  "/dev/sda1 / ext4 rw,relatime 0 0",
  "overlay /var/lib/docker overlay rw 0 0",
  "host_mnt /workspaces 9p rw,trans=virtio 0 0",
  "//server/share /mnt/smb cifs rw 0 0",
].join("\n");

test("parseMountFsType: 取最長前綴掛載點的 fstype", () => {
  assert.equal(parseMountFsType(MOUNTS, "/workspaces/janitarr/openspec"), "9p");
  assert.equal(parseMountFsType(MOUNTS, "/mnt/smb/x"), "cifs");
  assert.equal(parseMountFsType(MOUNTS, "/home/user/proj/openspec"), "ext4");
});

test("parseMountFsType: 掛載點完全相等也算涵蓋", () => {
  assert.equal(parseMountFsType(MOUNTS, "/workspaces"), "9p");
});

test("parseMountFsType: 處理掛載點中的八進位跳脫空白", () => {
  const mounts = "/dev/sda1 / ext4 rw 0 0\nhost /mnt/My\\040Files 9p rw 0 0";
  assert.equal(parseMountFsType(mounts, "/mnt/My Files/openspec"), "9p");
});

test("parseMountFsType: 無匹配（不應發生，root 通常涵蓋）回 null", () => {
  assert.equal(parseMountFsType("garbage line\n", "/x"), null);
});

// --- parsePollingOverride ---

test("parsePollingOverride: SPEK_WATCH_POLLING 真值/假值", () => {
  assert.equal(parsePollingOverride({ SPEK_WATCH_POLLING: "on" }), true);
  assert.equal(parsePollingOverride({ SPEK_WATCH_POLLING: "1" }), true);
  assert.equal(parsePollingOverride({ SPEK_WATCH_POLLING: "true" }), true);
  assert.equal(parsePollingOverride({ SPEK_WATCH_POLLING: "off" }), false);
  assert.equal(parsePollingOverride({ SPEK_WATCH_POLLING: "0" }), false);
});

test("parsePollingOverride: 退回 CHOKIDAR_USEPOLLING；未設定回 null", () => {
  assert.equal(parsePollingOverride({ CHOKIDAR_USEPOLLING: "true" }), true);
  assert.equal(parsePollingOverride({}), null);
  assert.equal(parsePollingOverride({ SPEK_WATCH_POLLING: "maybe" }), null);
});

test("parsePollingOverride: SPEK_WATCH_POLLING 優先於 CHOKIDAR_USEPOLLING", () => {
  assert.equal(
    parsePollingOverride({ SPEK_WATCH_POLLING: "off", CHOKIDAR_USEPOLLING: "true" }),
    false,
  );
});

// --- hasRemoteEnvIndicator ---

test("hasRemoteEnvIndicator: 偵測常見 remote / container 環境變數", () => {
  assert.equal(hasRemoteEnvIndicator({ REMOTE_CONTAINERS: "true" }), true);
  assert.equal(hasRemoteEnvIndicator({ CODESPACES: "true" }), true);
  assert.equal(hasRemoteEnvIndicator({ WSL_DISTRO_NAME: "Ubuntu" }), true);
  assert.equal(hasRemoteEnvIndicator({}), false);
});

// --- decidePolling ---

test("decidePolling: 明確覆寫優先於一切", () => {
  assert.equal(
    decidePolling({ override: true, platform: "darwin", fsType: "ext4", remoteIndicator: false }),
    true,
  );
  assert.equal(
    decidePolling({ override: false, platform: "linux", fsType: "9p", remoteIndicator: true }),
    false,
  );
});

test("decidePolling: 非 Linux 一律原生事件（false）", () => {
  assert.equal(
    decidePolling({ override: null, platform: "win32", fsType: null, remoteIndicator: true }),
    false,
  );
  assert.equal(
    decidePolling({ override: null, platform: "darwin", fsType: null, remoteIndicator: true }),
    false,
  );
});

test("decidePolling: Linux 上非事件 fstype → polling", () => {
  assert.equal(
    decidePolling({ override: null, platform: "linux", fsType: "9p", remoteIndicator: false }),
    true,
  );
});

test("decidePolling: Linux 上原生 fstype → 不 polling（即使環境像 container）", () => {
  assert.equal(
    decidePolling({ override: null, platform: "linux", fsType: "overlay", remoteIndicator: true }),
    false,
  );
});

test("decidePolling: fstype 無法判定時用 remoteIndicator 保底", () => {
  assert.equal(
    decidePolling({ override: null, platform: "linux", fsType: null, remoteIndicator: true }),
    true,
  );
  assert.equal(
    decidePolling({ override: null, platform: "linux", fsType: null, remoteIndicator: false }),
    false,
  );
});

// --- pollingInterval ---

test("pollingInterval: 預設 1000ms；CHOKIDAR_INTERVAL 覆寫；無效值退回預設", () => {
  assert.equal(pollingInterval({}), 1000);
  assert.equal(pollingInterval({ CHOKIDAR_INTERVAL: "500" }), 500);
  assert.equal(pollingInterval({ CHOKIDAR_INTERVAL: "abc" }), 1000);
  assert.equal(pollingInterval({ CHOKIDAR_INTERVAL: "-5" }), 1000);
  assert.equal(pollingInterval({ CHOKIDAR_INTERVAL: "0" }), 1000);
});

// --- shouldUsePolling ---

test("shouldUsePolling: 明確覆寫短路，回傳覆寫值且與路徑無關", () => {
  const p = "/definitely/not/a/real/path/openspec";
  assert.equal(shouldUsePolling(p, { env: { SPEK_WATCH_POLLING: "on" } }), true);
  assert.equal(shouldUsePolling(p, { env: { SPEK_WATCH_POLLING: "off" } }), false);
  assert.equal(shouldUsePolling(p, { env: { CHOKIDAR_USEPOLLING: "1" } }), true);
  // 短路 bug 會讓無覆寫時直接回傳 null 覆寫值；斷言型別可捕捉，且不受平台 / 掛載影響
  assert.equal(typeof shouldUsePolling(p, { env: {} }), "boolean");
});

test("shouldUsePolling: 覆寫優先於 extraRemoteIndicator 保底", () => {
  const p = "/definitely/not/a/real/path/openspec";
  assert.equal(
    shouldUsePolling(p, { env: { SPEK_WATCH_POLLING: "off" }, extraRemoteIndicator: true }),
    false,
  );
});

// --- withAuthoritativeChokidarEnv ---

test("withAuthoritativeChokidarEnv: callback 執行期間 env 對齊到權威值", () => {
  const env: NodeJS.ProcessEnv = { CHOKIDAR_USEPOLLING: "true", CHOKIDAR_INTERVAL: "0" };
  let seenUse: string | undefined;
  let seenInterval: string | undefined;
  withAuthoritativeChokidarEnv(
    false,
    1000,
    () => {
      seenUse = env.CHOKIDAR_USEPOLLING;
      seenInterval = env.CHOKIDAR_INTERVAL;
    },
    env,
  );
  assert.equal(seenUse, "false");
  assert.equal(seenInterval, "1000");
});

test("withAuthoritativeChokidarEnv: usePolling=true 時 env 設為字串 \"true\"", () => {
  const env: NodeJS.ProcessEnv = { CHOKIDAR_USEPOLLING: "false" };
  let seenUse: string | undefined;
  withAuthoritativeChokidarEnv(
    true,
    250,
    () => {
      seenUse = env.CHOKIDAR_USEPOLLING;
    },
    env,
  );
  assert.equal(seenUse, "true");
});

test("withAuthoritativeChokidarEnv: 結束後還原原有 env 值", () => {
  const env: NodeJS.ProcessEnv = { CHOKIDAR_USEPOLLING: "true", CHOKIDAR_INTERVAL: "250" };
  withAuthoritativeChokidarEnv(true, 500, () => {}, env);
  assert.equal(env.CHOKIDAR_USEPOLLING, "true");
  assert.equal(env.CHOKIDAR_INTERVAL, "250");
});

test("withAuthoritativeChokidarEnv: 原本未設定時結束後刪除（不留 undefined 字串）", () => {
  const env: NodeJS.ProcessEnv = {};
  withAuthoritativeChokidarEnv(true, 1000, () => {}, env);
  assert.equal("CHOKIDAR_USEPOLLING" in env, false);
  assert.equal("CHOKIDAR_INTERVAL" in env, false);
});

test("withAuthoritativeChokidarEnv: callback 拋錯仍還原 env 並向外拋出", () => {
  const env: NodeJS.ProcessEnv = { CHOKIDAR_USEPOLLING: "false" };
  assert.throws(() =>
    withAuthoritativeChokidarEnv(
      true,
      1000,
      () => {
        throw new Error("boom");
      },
      env,
    ),
  );
  assert.equal(env.CHOKIDAR_USEPOLLING, "false");
  assert.equal("CHOKIDAR_INTERVAL" in env, false);
});

test("withAuthoritativeChokidarEnv: 回傳 callback 的結果", () => {
  const env: NodeJS.ProcessEnv = {};
  const result = withAuthoritativeChokidarEnv(true, 1000, () => "watcher", env);
  assert.equal(result, "watcher");
});

File: packages/intellij/src/test/kotlin/com/spek/intellij/core/WatchPollingTest.kt
package com.spek.intellij.core

import java.io.File
import java.nio.file.Files
import kotlin.test.AfterTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue
import com.spek.intellij.core.WatchPolling.PollingSignals

/** 對齊 @spek/core watch-polling.test.ts 的判定邏輯測試 */
class WatchPollingTest {

    @Test
    fun fsTypeNeedsPolling_networkAndVirtualMounts() {
        for (t in listOf("9p", "v9fs", "drvfs", "cifs", "smb3", "nfs", "nfs4", "vboxsf", "prl_fs")) {
            assertTrue(WatchPolling.fsTypeNeedsPolling(t), t)
        }
    }

    @Test
    fun fsTypeNeedsPolling_fuseVariants() {
        assertTrue(WatchPolling.fsTypeNeedsPolling("fuse.sshfs"))
        assertTrue(WatchPolling.fsTypeNeedsPolling("fuse"))
        assertTrue(WatchPolling.fsTypeNeedsPolling("fuseblk"))
    }

    @Test
    fun fsTypeNeedsPolling_nativeFilesystems() {
        for (t in listOf("ext4", "overlay", "btrfs", "xfs", "tmpfs")) {
            assertFalse(WatchPolling.fsTypeNeedsPolling(t), t)
        }
    }

    @Test
    fun fsTypeNeedsPolling_caseInsensitiveAndNull() {
        assertTrue(WatchPolling.fsTypeNeedsPolling("9P"))
        assertTrue(WatchPolling.fsTypeNeedsPolling("NFS4"))
        assertFalse(WatchPolling.fsTypeNeedsPolling(null))
    }

    private val mounts = listOf(
        "rootfs / rootfs rw 0 0",
        "/dev/sda1 / ext4 rw,relatime 0 0",
        "overlay /var/lib/docker overlay rw 0 0",
        "host_mnt /workspaces 9p rw,trans=virtio 0 0",
        "//server/share /mnt/smb cifs rw 0 0",
    ).joinToString("\n")

    @Test
    fun parseMountFsType_longestPrefixWins() {
        assertEquals("9p", WatchPolling.parseMountFsType(mounts, "/workspaces/janitarr/openspec"))
        assertEquals("cifs", WatchPolling.parseMountFsType(mounts, "/mnt/smb/x"))
        assertEquals("ext4", WatchPolling.parseMountFsType(mounts, "/home/user/proj/openspec"))
    }

    @Test
    fun parseMountFsType_exactMountPointCovered() {
        assertEquals("9p", WatchPolling.parseMountFsType(mounts, "/workspaces"))
    }

    @Test
    fun parseMountFsType_handlesOctalEscapedSpace() {
        val m = "/dev/sda1 / ext4 rw 0 0\nhost /mnt/My\\040Files 9p rw 0 0"
        assertEquals("9p", WatchPolling.parseMountFsType(m, "/mnt/My Files/openspec"))
    }

    @Test
    fun parseMountFsType_noMatchReturnsNull() {
        assertNull(WatchPolling.parseMountFsType("garbage line\n", "/x"))
    }

    @Test
    fun parsePollingOverride_truthyFalsy() {
        assertEquals(true, WatchPolling.parsePollingOverride(mapOf("SPEK_WATCH_POLLING" to "on")))
        assertEquals(true, WatchPolling.parsePollingOverride(mapOf("SPEK_WATCH_POLLING" to "1")))
        assertEquals(false, WatchPolling.parsePollingOverride(mapOf("SPEK_WATCH_POLLING" to "off")))
        assertEquals(false, WatchPolling.parsePollingOverride(mapOf("SPEK_WATCH_POLLING" to "0")))
    }

    @Test
    fun parsePollingOverride_fallbackAndUnset() {
        assertEquals(true, WatchPolling.parsePollingOverride(mapOf("CHOKIDAR_USEPOLLING" to "true")))
        assertNull(WatchPolling.parsePollingOverride(emptyMap()))
        assertNull(WatchPolling.parsePollingOverride(mapOf("SPEK_WATCH_POLLING" to "maybe")))
    }

    @Test
    fun parsePollingOverride_spekWins() {
        assertEquals(
            false,
            WatchPolling.parsePollingOverride(
                mapOf("SPEK_WATCH_POLLING" to "off", "CHOKIDAR_USEPOLLING" to "true"),
            ),
        )
    }

    @Test
    fun hasRemoteEnvIndicator_detectsCommonVars() {
        assertTrue(WatchPolling.hasRemoteEnvIndicator(mapOf("REMOTE_CONTAINERS" to "true")))
        assertTrue(WatchPolling.hasRemoteEnvIndicator(mapOf("CODESPACES" to "true")))
        assertTrue(WatchPolling.hasRemoteEnvIndicator(mapOf("WSL_DISTRO_NAME" to "Ubuntu")))
        assertFalse(WatchPolling.hasRemoteEnvIndicator(emptyMap()))
    }

    @Test
    fun decidePolling_overrideWins() {
        assertTrue(WatchPolling.decidePolling(PollingSignals(true, false, "ext4", false)))
        assertFalse(WatchPolling.decidePolling(PollingSignals(false, true, "9p", true)))
    }

    @Test
    fun decidePolling_nonLinuxNeverPolls() {
        assertFalse(WatchPolling.decidePolling(PollingSignals(null, false, null, true)))
    }

    @Test
    fun decidePolling_linuxNonEventFsPolls() {
        assertTrue(WatchPolling.decidePolling(PollingSignals(null, true, "9p", false)))
    }

    @Test
    fun decidePolling_linuxNativeFsNoPoll() {
        assertFalse(WatchPolling.decidePolling(PollingSignals(null, true, "overlay", true)))
    }

    @Test
    fun decidePolling_unknownFsFallsBackToRemoteIndicator() {
        assertTrue(WatchPolling.decidePolling(PollingSignals(null, true, null, true)))
        assertFalse(WatchPolling.decidePolling(PollingSignals(null, true, null, false)))
    }

    @Test
    fun pollingIntervalMs_defaultAndOverride() {
        assertEquals(1000L, WatchPolling.pollingIntervalMs(emptyMap()))
        assertEquals(500L, WatchPolling.pollingIntervalMs(mapOf("CHOKIDAR_INTERVAL" to "500")))
        assertEquals(1000L, WatchPolling.pollingIntervalMs(mapOf("CHOKIDAR_INTERVAL" to "abc")))
        assertEquals(1000L, WatchPolling.pollingIntervalMs(mapOf("CHOKIDAR_INTERVAL" to "-5")))
    }

    // --- scanSnapshot：輪詢偵測機制（對應 chokidar 的 polling smoke test） ---

    private val tempDirs = mutableListOf<File>()

    @AfterTest
    fun cleanupTempDirs() {
        tempDirs.forEach { it.deleteRecursively() }
    }

    private fun mkTempOpenspec(): File {
        val dir = Files.createTempDirectory("spek-poll-").toFile()
        tempDirs.add(dir)
        return dir
    }

    @Test
    fun scanSnapshot_detectsNewlyCreatedFile() {
        val dir = mkTempOpenspec()
        val before = WatchPolling.scanSnapshot(dir)
        // watcher ready 後才新建的檔案（容器內失效的情境）
        File(dir, "brainstorm.md").writeText("# hi")
        val after = WatchPolling.scanSnapshot(dir)
        assertNotEquals(before, after, "新建 .md 後快照應改變")
        assertTrue(after.keys.any { it.endsWith("brainstorm.md") })
    }

    @Test
    fun scanSnapshot_detectsFileInNewlyCreatedNestedDir() {
        val dir = mkTempOpenspec()
        val before = WatchPolling.scanSnapshot(dir)
        val nested = File(dir, "changes/x/specs/topic")
        nested.mkdirs()
        File(nested, "spec.md").writeText("# spec")
        assertNotEquals(before, WatchPolling.scanSnapshot(dir), "巢狀新建目錄內的 spec.md 應被偵測")
    }

    @Test
    fun scanSnapshot_detectsSizeChangeWithUnchangedMtime() {
        val dir = mkTempOpenspec()
        val f = File(dir, "spec.md")
        f.writeText("# hi")
        // 釘住 mtime 到固定整秒，模擬 9p / NFS 上就地 append 時 mtime 不進位的情境
        val fixedMtime = 1_700_000_000_000L
        f.setLastModified(fixedMtime)
        val before = WatchPolling.scanSnapshot(dir)
        f.appendText(" appended, size grew")
        f.setLastModified(fixedMtime)
        val after = WatchPolling.scanSnapshot(dir)
        assertNotEquals(before, after, "mtime 不變但 size 改變時仍應偵測到（快照值需納入 size）")
    }

    @Test
    fun scanSnapshot_onlyMdAndYaml_skipsDotDirs() {
        val dir = mkTempOpenspec()
        File(dir, "keep.md").writeText("a")
        File(dir, "keep.yaml").writeText("b")
        File(dir, "ignore.txt").writeText("c")
        File(dir, ".git").apply { mkdirs() }.let { File(it, "inside.md").writeText("d") }
        val snap = WatchPolling.scanSnapshot(dir)
        assertEquals(2, snap.size, "只計入 .md/.yaml 且跳過 dotdir")
        assertTrue(snap.keys.any { it.endsWith("keep.md") })
        assertTrue(snap.keys.any { it.endsWith("keep.yaml") })
    }
}
